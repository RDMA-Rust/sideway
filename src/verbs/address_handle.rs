use rdma_mummy_sys::ibv_gid;
use std::{
    fmt,
    net::Ipv6Addr,
};

#[derive(Default, Clone, Copy, Debug)]
pub struct Gid {
    pub raw: [u8; 16],
}

impl AsMut<ibv_gid> for Gid {
    fn as_mut(&mut self) -> &mut ibv_gid {
        unsafe { &mut *self.raw.as_mut_ptr().cast::<ibv_gid>() }
    }
}

impl From<ibv_gid> for Gid {
    fn from(gid: ibv_gid) -> Self {
        Self {
            raw: unsafe { gid.raw },
        }
    }
}

impl From<Gid> for ibv_gid {
    fn from(mut gid: Gid) -> Self {
        *gid.as_mut()
    }
}

impl From<Gid> for Ipv6Addr {
    fn from(gid: Gid) -> Self {
        Ipv6Addr::from(gid.raw)
    }
}

impl fmt::Display for Gid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for (i, &byte) in self.raw.iter().enumerate() {
            if i > 0 && i % 2 == 0 {
                write!(f, ":")?;
            }
            write!(f, "{:02x}", byte)?;
        }
        Ok(())
    }
}

impl Gid {
    pub fn is_zero(&self) -> bool {
        let (prefix, aligned, suffix) = unsafe { self.raw.align_to::<u128>() };

        prefix.iter().all(|&x| x == 0) && suffix.iter().all(|&x| x == 0) && aligned.iter().all(|&x| x == 0)
    }
}
