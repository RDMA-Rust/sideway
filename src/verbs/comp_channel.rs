use std::{marker::PhantomData, ptr};

use rdma_mummy_sys::{ibv_comp_channel, ibv_create_comp_channel, ibv_destroy_comp_channel};

use super::rdma_context::RdmaContext;

#[derive(Debug)]
pub struct CompletionChannel<'ctx> {
    pub(crate) channel: ptr::NonNull<ibv_comp_channel>,
    // phantom data for device context
    _dev_ctx: PhantomData<&'ctx RdmaContext>,
}

impl<'ctx> Drop for CompletionChannel<'ctx> {
    fn drop(&mut self) {
        let ret = unsafe { ibv_destroy_comp_channel(self.channel.as_ptr()) };
        assert_eq!(ret, 0);
    }
}

impl<'ctx> CompletionChannel<'ctx> {
    pub fn new<'a>(dev_ctx: &'a RdmaContext) -> Result<CompletionChannel<'a>, String> {
        let comp_channel = unsafe { ibv_create_comp_channel(dev_ctx.context) };
        Ok(CompletionChannel {
            channel: ptr::NonNull::new(comp_channel).ok_or(String::from("ibv_create_comp_channel failed"))?,
            _dev_ctx: PhantomData,
        })
    }
}
