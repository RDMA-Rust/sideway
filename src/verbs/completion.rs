use std::num::NonZeroU32;
use std::os::raw::c_void;
use std::ptr;
use std::ptr::NonNull;
use std::{marker::PhantomData, mem::MaybeUninit};

use super::device_context::DeviceContext;
use rdma_mummy_sys::{
    ibv_comp_channel, ibv_cq, ibv_cq_ex, ibv_cq_init_attr_ex, ibv_create_comp_channel, ibv_create_cq, ibv_create_cq_ex,
    ibv_destroy_comp_channel, ibv_destroy_cq, ibv_end_poll, ibv_next_poll, ibv_pd, ibv_poll_cq, ibv_poll_cq_attr,
    ibv_start_poll, ibv_wc, ibv_wc_read_byte_len, ibv_wc_read_completion_ts, ibv_wc_read_opcode,
    ibv_wc_read_vendor_err,
};

#[derive(Debug)]
pub struct CompletionChannel<'res> {
    pub(crate) channel: NonNull<ibv_comp_channel>,
    // phantom data for device context
    _phantom: PhantomData<&'res ()>,
}

impl Drop for CompletionChannel<'_> {
    fn drop(&mut self) {
        let ret = unsafe { ibv_destroy_comp_channel(self.channel.as_ptr()) };
        assert_eq!(ret, 0);
    }
}

impl<'res> CompletionChannel<'res> {
    pub fn new<'ctx>(dev_ctx: &'ctx DeviceContext) -> Result<CompletionChannel<'res>, String>
    where
        'ctx: 'res,
    {
        let comp_channel = unsafe { ibv_create_comp_channel(dev_ctx.context) };
        Ok(CompletionChannel {
            channel: NonNull::new(comp_channel).ok_or(String::from("ibv_create_comp_channel failed"))?,
            _phantom: PhantomData,
        })
    }
}

pub trait CompletionQueue {
    /// # Safety
    ///
    /// return the basic handle of CQ;
    /// we mark this method unsafe because the lifetime of ibv_cq is not
    /// associated with the return value.
    unsafe fn cq(&self) -> NonNull<ibv_cq>;
}

#[derive(Debug)]
pub struct BasicCompletionQueue<'res> {
    pub(crate) cq: NonNull<ibv_cq>,
    poll_batch: NonZeroU32,
    // phantom data for dev_ctx & comp_channel
    _phantom: PhantomData<&'res ()>,
}

impl Drop for BasicCompletionQueue<'_> {
    fn drop(&mut self) {
        let ret = unsafe { ibv_destroy_cq(self.cq.as_ptr()) };
        assert_eq!(ret, 0);
    }
}

impl CompletionQueue for BasicCompletionQueue<'_> {
    unsafe fn cq(&self) -> NonNull<ibv_cq> {
        self.cq
    }
}

impl BasicCompletionQueue<'_> {
    pub fn start_poll(&self) -> Result<BasicPoller<'_>, String> {
        let mut cqes = Vec::<ibv_wc>::with_capacity(self.poll_batch.get() as _);

        let ret = unsafe {
            ibv_poll_cq(
                self.cq.as_ptr(),
                self.poll_batch.get().try_into().unwrap(),
                cqes.as_mut_ptr(),
            )
        };

        unsafe {
            match ret {
                0 => Err("no valid cqes".to_string()),
                err if err < 0 => Err(format!("ibv_poll_cq failed, ret={err}")),
                res => Ok(BasicPoller {
                    cq: self.cq(),
                    wcs: {
                        cqes.set_len(res as _);
                        cqes
                    },
                    status: if res < self.poll_batch.get().try_into().unwrap_unchecked() {
                        BasicCompletionQueueState::Drained
                    } else {
                        BasicCompletionQueueState::Ready
                    },
                    current: 0,
                    _phantom: PhantomData,
                }),
            }
        }
    }

    pub fn setup_poll_batch(&mut self, size: NonZeroU32) {
        self.poll_batch = size;
    }
}

#[derive(Debug)]
pub struct ExtendedCompletionQueue<'res> {
    pub(crate) cq_ex: NonNull<ibv_cq_ex>,
    // phantom data for dev_ctx & comp_channel
    _phantom: PhantomData<&'res ()>,
}

impl Drop for ExtendedCompletionQueue<'_> {
    fn drop(&mut self) {
        // TODO convert cq_ex to cq (port ibv_cq_ex_to_cq in rdma-mummy-sys)
        let ret = unsafe { ibv_destroy_cq(self.cq_ex.as_ptr().cast()) };
        assert_eq!(ret, 0);
    }
}

impl CompletionQueue for ExtendedCompletionQueue<'_> {
    unsafe fn cq(&self) -> NonNull<ibv_cq> {
        // TODO convert cq_ex to cq (port ibv_cq_ex_to_cq in rdma-mummy-sys)
        self.cq_ex.cast()
    }
}

impl ExtendedCompletionQueue<'_> {
    pub fn start_poll(&self) -> Result<ExtendedPoller<'_>, String> {
        let ret = unsafe {
            ibv_start_poll(
                self.cq_ex.as_ptr(),
                MaybeUninit::<ibv_poll_cq_attr>::zeroed().as_mut_ptr(),
            )
        };

        match ret {
            0 => Ok(ExtendedPoller {
                cq: self.cq_ex,
                is_first: true,
                _phantom: PhantomData,
            }),
            err => Err(format!("ibv_start_poll failed, ret={err}")),
        }
    }
}

// generic builder for both cq and cq_ex
pub struct CompletionQueueBuilder<'res> {
    dev_ctx: &'res DeviceContext,
    init_attr: ibv_cq_init_attr_ex,
}

impl<'res> CompletionQueueBuilder<'res> {
    pub fn new<'ctx>(dev_ctx: &'ctx DeviceContext) -> Self
    where
        'ctx: 'res,
    {
        // set default params for init_attr
        CompletionQueueBuilder {
            dev_ctx,
            init_attr: ibv_cq_init_attr_ex {
                cqe: 1024,
                cq_context: ptr::null_mut::<c_void>(),
                channel: ptr::null_mut::<ibv_comp_channel>(),
                comp_vector: 0,
                // TODO(zhp): setup default flags for CQ
                wc_flags: 0,
                comp_mask: 0,
                flags: 0,
                parent_domain: ptr::null_mut::<ibv_pd>(),
            },
        }
    }

    pub fn setup_cqe(&mut self, cqe: u32) -> &mut Self {
        self.init_attr.cqe = cqe;
        self
    }

    pub fn setup_cq_context(&mut self, cq_context: *mut c_void) -> &mut Self {
        self.init_attr.cq_context = cq_context;
        self
    }

    pub fn setup_comp_channel<'channel>(&mut self, channel: &'channel CompletionChannel, comp_vector: u32) -> &mut Self
    where
        'channel: 'res,
    {
        self.init_attr.channel = channel.channel.as_ptr();
        self.init_attr.comp_vector = comp_vector;
        self
    }

    // TODO(fuji): set various attributes

    // build extended cq
    pub fn build_ex(&self) -> Result<ExtendedCompletionQueue<'res>, String> {
        // create a copy of init_attr since ibv_create_cq_ex requires a mutable pointer to it
        let mut init_attr = self.init_attr;

        let cq_ex = unsafe { ibv_create_cq_ex(self.dev_ctx.context, &mut init_attr as *mut _) };

        Ok(ExtendedCompletionQueue {
            cq_ex: NonNull::new(cq_ex).ok_or(String::from("ibv_create_cq_ex failed"))?,
            // associate the lifetime of dev_ctx & comp_channel with CQ
            _phantom: PhantomData,
        })
    }

    // build basic cq
    pub fn build(&self) -> Result<BasicCompletionQueue<'res>, String> {
        let cq = unsafe {
            ibv_create_cq(
                self.dev_ctx.context,
                self.init_attr.cqe as _,
                self.init_attr.cq_context,
                self.init_attr.channel,
                self.init_attr.comp_vector as _,
            )
        };
        Ok(BasicCompletionQueue {
            cq: NonNull::new(cq).ok_or(String::from("ibv_create_cq failed"))?,
            poll_batch: unsafe { NonZeroU32::new(32).unwrap_unchecked() },
            // associate the lifetime of dev_ctx & comp_channel with CQ
            _phantom: PhantomData,
        })
    }
}

// TODO trait for both cq and cq_ex?

pub struct BasicWorkCompletion<'iter> {
    wc: ibv_wc,
    _phantom: PhantomData<&'iter ()>,
}

impl<'iter> BasicWorkCompletion<'iter> {
    pub fn wr_id(&self) -> u64 {
        self.wc.wr_id
    }

    pub fn status(&self) -> u32 {
        self.wc.status
    }

    pub fn opcode(&self) -> u32 {
        self.wc.opcode
    }
}

pub struct ExtendedWorkCompletion<'iter> {
    cq: NonNull<ibv_cq_ex>,
    _phantom: PhantomData<&'iter ()>,
}

impl<'iter> ExtendedWorkCompletion<'iter> {
    pub fn wr_id(&self) -> u64 {
        unsafe { self.cq.as_ref().wr_id }
    }

    pub fn status(&self) -> u32 {
        unsafe { self.cq.as_ref().status }
    }

    pub fn opcode(&self) -> u32 {
        unsafe { ibv_wc_read_opcode(self.cq.as_ptr()) }
    }

    pub fn vendor_err(&self) -> u32 {
        unsafe { ibv_wc_read_vendor_err(self.cq.as_ptr()) }
    }

    pub fn byte_len(&self) -> u32 {
        unsafe { ibv_wc_read_byte_len(self.cq.as_ptr()) }
    }

    pub fn completion_timestamp(&self) -> u64 {
        unsafe { ibv_wc_read_completion_ts(self.cq.as_ptr()) }
    }
}

#[derive(PartialEq, Eq)]
enum BasicCompletionQueueState {
    // Ready means Completion Queue still has elements waiting to be polled.
    // For Basic CQ, it means we are ready to return a work completion, if no more
    // completions are there, we would call `ibv_poll_cq` immediately to get more.
    Ready,
    // Drained means the CQ is drained in this round of polling, we would return
    // None after user consuming the last work completion, if user want to use
    // `next()` to get more work completions, it would trigger `ibv_poll_cq` calls.
    Drained,
    // Empty means CQ is *likely* to be empty, if user want to get next work
    // completion, we would call `ibv_poll_cq` to check if there is a new one.
    Empty,
}

pub struct BasicPoller<'cq> {
    cq: NonNull<ibv_cq>,
    wcs: Vec<ibv_wc>,
    status: BasicCompletionQueueState,
    current: usize,
    _phantom: PhantomData<&'cq ()>,
}

impl<'cq> Iterator for BasicPoller<'cq> {
    type Item = BasicWorkCompletion<'cq>;

    fn next(&mut self) -> Option<Self::Item> {
        use BasicCompletionQueueState::*;

        let current = self.current;
        let len = self.wcs.len();

        if (self.status == Ready || self.status == Drained) && current < len {
            let wc = unsafe {
                BasicWorkCompletion {
                    wc: *self.wcs.get_unchecked(current),
                    _phantom: PhantomData,
                }
            };
            self.current += 1;
            return Some(wc);
        }

        if self.status == Drained && current >= len {
            // Completion Queue has been drained once, return None to let user
            // get sense of it.
            self.status = Empty;
            return None;
        }

        // Status is Ready, but all work completions have been consumed, poll
        // one more time to get more work completions.
        // Or status is Empty, try if we could get new work completions.
        let ret = unsafe {
            ibv_poll_cq(
                self.cq.as_ptr(),
                self.wcs.capacity().try_into().unwrap_unchecked(),
                self.wcs.as_mut_ptr(),
            )
        };

        if ret > 0 {
            unsafe {
                if ret < self.wcs.capacity().try_into().unwrap_unchecked() {
                    self.status = Drained;
                } else {
                    self.status = Ready;
                }

                self.wcs.set_len(ret as usize);
            }
            let wc = unsafe {
                BasicWorkCompletion {
                    wc: *self.wcs.get_unchecked(0),
                    _phantom: PhantomData,
                }
            };
            self.current = 1;
            Some(wc)
        } else {
            self.status = Empty;
            None
        }
    }
}

pub struct ExtendedPoller<'cq> {
    cq: NonNull<ibv_cq_ex>,
    is_first: bool,
    _phantom: PhantomData<&'cq ()>,
}

impl Drop for ExtendedPoller<'_> {
    fn drop(&mut self) {
        unsafe { ibv_end_poll(self.cq.as_ptr()) }
    }
}

impl<'cq> Iterator for ExtendedPoller<'cq> {
    type Item = ExtendedWorkCompletion<'cq>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.is_first {
            self.is_first = false;
            Some(ExtendedWorkCompletion {
                cq: self.cq,
                _phantom: PhantomData,
            })
        } else {
            let ret = unsafe { ibv_next_poll(self.cq.as_ptr()) };

            if ret != 0 {
                None
            } else {
                Some(ExtendedWorkCompletion {
                    cq: self.cq,
                    _phantom: PhantomData,
                })
            }
        }
    }
}
