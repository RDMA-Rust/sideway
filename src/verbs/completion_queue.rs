use std::os::raw::c_void;
use std::ptr;
use std::{
    alloc::{alloc_zeroed, dealloc, Layout},
    marker::PhantomData,
};

use super::{protection_domain::ProtectionDomain, rdma_context::RdmaContext};
use rdma_mummy_sys::{
    ibv_comp_channel, ibv_cq, ibv_cq_ex, ibv_cq_init_attr_ex, ibv_create_cq_ex, ibv_destroy_cq, ibv_pd,
};

#[derive(Debug)]
pub struct CqInitAttr {
    pub(crate) attr_ptr: *mut ibv_cq_init_attr_ex,
}

impl CqInitAttr {
    pub fn new() -> Self {
        let attr_ptr = unsafe { alloc_zeroed(Layout::new::<ibv_cq_init_attr_ex>()) as *mut ibv_cq_init_attr_ex };
        CqInitAttr { attr_ptr }
    }

    pub fn set_cqe(self, cqe: u32) -> Self {
        unsafe {
            (*self.attr_ptr).cqe = cqe;
        }
        self
    }

    pub fn set_pd(self, pd: &ProtectionDomain) -> Self {
        unsafe {
            (*self.attr_ptr).parent_domain = pd.pd_ptr;
        }
        self
    }
}

impl Drop for CqInitAttr {
    fn drop(&mut self) {
        unsafe {
            dealloc(self.attr_ptr as _, Layout::new::<ibv_cq_init_attr_ex>());
        }
    }
}

#[derive(Debug)]
pub struct CompletionQueue {
    pub(crate) cq_ptr: *mut ibv_cq,
}

impl Drop for CompletionQueue {
    fn drop(&mut self) {
        unsafe {
            ibv_destroy_cq(self.cq_ptr);
        }
    }
}

#[derive(Debug)]
pub struct CompletionQueueEx {
    pub(crate) cq_ptr: *mut ibv_cq_ex,
}

#[derive(Debug)]
pub struct ExtendedCompletionQueue<'ctx> {
    pub(crate) cq_ex: ptr::NonNull<ibv_cq_ex>,
    // phantom data for device context
    _dev_ctx: PhantomData<&'ctx ()>,
    // TODO(zhp): comp_channel, lifetime
}

impl<'ctx> Drop for ExtendedCompletionQueue<'ctx> {
    fn drop(&mut self) {
        // TODO convert cq_ex to cq (port ibv_cq_ex_to_cq in rdma-mummy-sys)
        let ret = unsafe { ibv_destroy_cq(self.cq_ex.as_ptr().cast()) };
        assert_eq!(ret, 0);
    }
}

pub struct ExtendedCompletionQueueBuilder<'ctx> {
    dev_ctx: &'ctx RdmaContext,
    init_attr: ibv_cq_init_attr_ex,
}

impl<'ctx> ExtendedCompletionQueueBuilder<'ctx> {
    pub fn new<'a>(dev_ctx: &'a RdmaContext) -> ExtendedCompletionQueueBuilder<'a> {
        // set default params for init_attr
        ExtendedCompletionQueueBuilder {
            dev_ctx,
            init_attr: ibv_cq_init_attr_ex {
                cqe: 1024,
                cq_context: ptr::null::<c_void>() as *mut _,
                channel: ptr::null::<ibv_comp_channel>() as *mut _,
                comp_vector: 0,
                // TODO(zhp): setup default flags for CQ
                wc_flags: 0,
                comp_mask: 0,
                flags: 0,
                parent_domain: ptr::null::<ibv_pd>() as *mut _,
            },
        }
    }
    // TODO(zhp): setup comp_channel?

    // TODO(zhp): new, set various attributes
    pub fn setup_cqe(&mut self, cqe: u32) -> &mut Self {
        self.init_attr.cqe = cqe;
        self
    }

    pub fn build(&self) -> Result<ExtendedCompletionQueue<'ctx>, String> {
        // create a copy of init_attr since ibv_create_cq_ex requires a mutable pointer to it
        let mut init_attr = self.init_attr.clone();
        match unsafe { ibv_create_cq_ex(self.dev_ctx.context, &mut init_attr as *mut _) } {
            Some(cq_ex) => Ok(ExtendedCompletionQueue {
                cq_ex: ptr::NonNull::new(cq_ex).ok_or(String::from("ibv_create_cq_ex failed"))?,
                // TODO should associate the lifetime of dev_ctx with CQ
                _dev_ctx: PhantomData,
            }),
            None => Err(String::from("ibv_create_cq_ex failed")),
        }
    }
}

// TODO trait for both cq and cq_ex?
