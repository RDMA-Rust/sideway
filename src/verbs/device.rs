use std::{ffi::CStr, io, marker::PhantomData};

use rdma_mummy_sys::{
    ibv_device, ibv_free_device_list, ibv_get_device_guid, ibv_get_device_list, ibv_get_device_name, ibv_open_device,
};

use crate::verbs::device_context::DeviceContext;

pub struct DeviceList {
    devices: *mut *mut ibv_device,
    num_devices: usize,
}

impl DeviceList {
    pub fn new() -> io::Result<DeviceList> {
        let mut num_devices: i32 = 0;
        let devices = unsafe { ibv_get_device_list(&mut num_devices as *mut _) };
        if devices.is_null() {
            return Err(io::Error::last_os_error());
        }

        Ok(DeviceList {
            devices,
            num_devices: num_devices as usize,
        })
    }

    pub fn iter<'list>(&'list self) -> DeviceListIter<'list> {
        DeviceListIter {
            current: 0,
            total: self.num_devices,
            devices: self,
        }
    }

    pub fn len(&self) -> usize {
        self.num_devices
    }
}

impl Drop for DeviceList {
    fn drop(&mut self) {
        unsafe { ibv_free_device_list(self.devices) };
    }
}

impl<'list> IntoIterator for &'list DeviceList {
    type Item = <DeviceListIter<'list> as Iterator>::Item;
    type IntoIter = DeviceListIter<'list>;
    fn into_iter(self) -> Self::IntoIter {
        DeviceListIter {
            current: 0,
            total: self.num_devices,
            devices: self,
        }
    }
}

pub struct DeviceListIter<'list> {
    current: usize,
    total: usize,
    devices: &'list DeviceList,
}

impl<'list> Iterator for DeviceListIter<'list> {
    type Item = Device<'list>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current >= self.total {
            None
        } else {
            unsafe {
                let device = *self.devices.devices.add(self.current);
                self.current += 1;
                Some(Device::new(device, self.devices))
            }
        }
    }
}

pub struct Device<'list> {
    device: *mut ibv_device,
    _dev_list: PhantomData<&'list ()>,
}

impl Device<'_> {
    pub fn new<'list>(device: *mut ibv_device, _devices: &'list DeviceList) -> Self {
        Device {
            device,
            _dev_list: PhantomData,
        }
    }

    pub fn open(&self) -> io::Result<DeviceContext> {
        unsafe {
            let context = ibv_open_device(self.device);
            if context.is_null() {
                return Err(io::Error::last_os_error());
            }
            Ok(DeviceContext { context })
        }
    }

    pub fn name(&self) -> Option<String> {
        unsafe {
            let name = ibv_get_device_name(self.device);
            if name.is_null() {
                None
            } else {
                Some(String::from_utf8_lossy(CStr::from_ptr(name).to_bytes()).to_string())
            }
        }
    }

    pub fn guid(&self) -> u64 {
        unsafe { ibv_get_device_guid(self.device) }
    }
}
