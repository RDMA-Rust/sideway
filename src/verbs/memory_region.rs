use super::protection_domain::ProtectionDomain;
use rdma_mummy_sys::{ibv_dereg_mr, ibv_mr};
use std::{
    alloc::{dealloc, handle_alloc_error, Layout},
    marker::PhantomData,
    ptr::NonNull,
};

// This buffer definition is from arrow-buffer
#[derive(Clone, Debug)]
pub struct Buffer {
    pub data: NonNull<u8>,
    pub len: usize,
    layout: Layout,
}

impl Buffer {
    #[inline]
    pub fn from_len_zeroed(len: usize) -> Self {
        let layout = Layout::from_size_align(len, 8).unwrap();
        let data = match layout.size() {
            0 => panic!(),
            _ => {
                let raw_ptr = unsafe { std::alloc::alloc_zeroed(layout) };
                NonNull::new(raw_ptr).unwrap_or_else(|| handle_alloc_error(layout))
            },
        };
        Self { data, len, layout }
    }
}

#[derive(Debug)]
pub struct MemoryRegion<'pd> {
    pub buf: Buffer,
    mr: NonNull<ibv_mr>,
    _pd: PhantomData<&'pd ()>,
}

impl Drop for MemoryRegion<'_> {
    fn drop(&mut self) {
        unsafe {
            ibv_dereg_mr(self.mr.as_mut());
            dealloc(self.buf.data.as_mut(), self.buf.layout);
        }
    }
}

impl MemoryRegion<'_> {
    pub(crate) fn new<'pd>(_pd: &'pd ProtectionDomain<'pd>, buf: Buffer, mr: NonNull<ibv_mr>) -> MemoryRegion<'pd> {
        MemoryRegion {
            buf,
            mr,
            _pd: PhantomData,
        }
    }

    pub fn lkey(&self) -> u32 {
        unsafe { self.mr.as_ref().lkey }
    }

    pub fn rkey(&self) -> u32 {
        unsafe { self.mr.as_ref().rkey }
    }
}
