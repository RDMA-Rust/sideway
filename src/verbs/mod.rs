pub mod address;
pub mod completion;
pub mod device;
pub mod device_context;
pub mod memory_region;
pub mod protection_domain;
pub mod queue_pair;
