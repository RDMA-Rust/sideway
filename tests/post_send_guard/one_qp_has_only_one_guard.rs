use sideway::verbs::{
    address::AddressHandleAttribute,
    device,
    device_context::Mtu,
    queue_pair::{PostSendGuard, QueuePair, QueuePairAttribute, QueuePairState},
    AccessFlags,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device_list = device::DeviceList::new()?;
    for device in &device_list {
        let ctx = device.open().unwrap();

        let pd = ctx.alloc_pd().unwrap();
        let _mr = pd.reg_managed_mr(64).unwrap();

        let _comp_channel = ctx.create_comp_channel().unwrap();
        let mut cq_builder = ctx.create_cq_builder();
        let sq = cq_builder.setup_cqe(128).build_ex().unwrap();
        let rq = cq_builder.setup_cqe(128).build_ex().unwrap();

        let mut builder = pd.create_qp_builder();

        // block for extended qp
        {
            let mut qp = builder
                .setup_max_inline_data(128)
                .setup_send_cq(&sq)
                .setup_recv_cq(&rq)
                .build_ex()
                .unwrap();

            println!("qp pointer is {:?}", qp);
            // modify QP to INIT state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::Init)
                .setup_pkey_index(0)
                .setup_port(1)
                .setup_access_flags(AccessFlags::LocalWrite | AccessFlags::RemoteWrite);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::Init, qp.state());

            // modify QP to RTR state, set dest qp as itself
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToReceive)
                .setup_path_mtu(Mtu::Mtu1024)
                .setup_dest_qp_num(qp.qp_number())
                .setup_rq_psn(1)
                .setup_max_dest_read_atomic(0)
                .setup_min_rnr_timer(0);
            // setup address vector
            let mut ah_attr = AddressHandleAttribute::new();
            let gid_entries = ctx.query_gid_table().unwrap();

            ah_attr
                .setup_dest_lid(1)
                .setup_port(1)
                .setup_service_level(1)
                .setup_grh_src_gid_index(gid_entries[0].gid_index().try_into().unwrap())
                .setup_grh_dest_gid(&gid_entries[0].gid())
                .setup_grh_hop_limit(64);
            attr.setup_address_vector(&ah_attr);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToReceive, qp.state());

            // modify QP to RTS state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToSend)
                .setup_sq_psn(1)
                .setup_timeout(12)
                .setup_retry_cnt(7)
                .setup_rnr_retry(7)
                .setup_max_read_atomic(0);

            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToSend, qp.state());

            let guard = qp.start_post_send();

            // while holding a post send guard, we can't build a post send guard at the same time
            let _guard_2 = qp.start_post_send();

            let _res = guard.post().unwrap();
        }
    }

    Ok(())
}
