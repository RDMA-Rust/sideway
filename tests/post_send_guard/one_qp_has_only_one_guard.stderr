error[E0499]: cannot borrow `qp` as mutable more than once at a time
  --> tests/post_send_guard/one_qp_has_only_one_guard.rs:92:28
   |
89 |             let guard = qp.start_post_send();
   |                         -- first mutable borrow occurs here
...
92 |             let _guard_2 = qp.start_post_send();
   |                            ^^ second mutable borrow occurs here
93 |
94 |             let _res = guard.post().unwrap();
   |                        ----- first borrow later used here

error[E0499]: cannot borrow `qp` as mutable more than once at a time
   --> tests/post_send_guard/one_qp_has_only_one_guard.rs:157:28
    |
154 |             let guard = qp.start_post_send();
    |                         -- first mutable borrow occurs here
...
157 |             let _guard_2 = qp.start_post_send();
    |                            ^^ second mutable borrow occurs here
158 |
159 |             let _res = guard.post().unwrap();
    |                        ----- first borrow later used here
